### PR TITLE
Feature/create note

### PR DESCRIPTION
**Pull request by DaniilVolobuiev**
## Describe your changes

> feat: Notes. Get and create notes. Ability to upload and download files. Ability to search and sort. Pagination.
> Skeleton. Loader for notes download.
> functions: debounce
> hooks: useHighLight hook to mark the searched text. 
> RTK query: Note Service

screenshot-1682842555389

screenshot-1682842794768

screenshot-1682842940466

screenshot-1682843020031

## Issue ticket code (and/or) and link

> [Link to JIRA ticket](https://smartdc23.atlassian.net/jira/software/projects/SMAR/boards/1/backlog?selectedIssue=SMAR-59) smar 59

### General

- [ ] Assigned myself to the PR
- [ ] Assigned the appropriate labels to the PR
- [ ] Assigned the appropriate reviewers to the PR
- [ ] Updated the documentation
- [ ] Performed a self-review of my code
- [ ] Types for input and output parameters
- [ ] Don't have "any" on my code
- [ ] Used the try/catch pattern for error handling
- [ ] Don't have magic numbers
- [ ] Compare only with constants not with strings
- [ ] No ternary operator inside ternary operator
- [ ] Don't have commented code
- [ ] no links in the code, env links should be in env file (for example: server url), constant links (for example default avatar URL) should be in constant file.
- [ ] Used camelCase for variables and functions
- [ ] Date and time formats are on the constants
- [ ] Functions is public only if it's used outside the class
- [ ] No hardcoded values
- [ ] covered by tests
- [ ] check your commit messages meet the [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/).

### Frontend

- [x] Components and business logic are separated
- [x] Colors, Font Size, and Font Name is on the theme or in the constants
- [x] No text in the components, use i18n approach
- [x] No inline styles
- [x] Imports are absolute
- [x] Attach a screenshot if PR has visual changes.

### CODE REVIEW

File: src/components/FileInput/index.tsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+          {...otherProps}
+        />
+      </FileUploadContainer>
+      <FilePreviewContainer>
+        <span> {t('Notes.ToUpload')}</span>
+        <PreviewList>
+          {Object.keys(files).map((fileName, index) => {
+            let previewFiles: { [key: string]: any } = { ...files };
+            let file = previewFiles[fileName];
+            let isImageFile = file.type.split('/')[0] === 'image';
```
Comment (lyudad) [inline src/components/FileInput/index.tsx]: no logic in render, move it somewhere

File: src/components/FileInput/styles.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  margin-top: 0;
+  margin-bottom: 20px;
+  text-align: center;
+`;
+
+export const UploadFileBtn = styled.button`
+  box-sizing: border-box;
+  appearance: none;
+  background-color: transparent;
+  border: 2px solid #3498db;
```
Comment (lyudad) [inline src/components/FileInput/styles.ts]: color

File: src/components/FileInput/styles.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  &:after {
+    content: '';
+    position: absolute;
+    display: block;
+    top: 0;
+    left: 50%;
+    transform: translateX(-50%);
+    width: 0;
+    height: 100%;
+    background: #3498db;
```
Comment (lyudad) [inline src/components/FileInput/styles.ts]: checklist

File: src/components/LoadMoreButton/index.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  const notesLocalLength = notesLocal.length;
+
+  const fetchMore = React.useCallback(() => {
+    dispatch(noteFilterActions.setSkipAmount());
+  }, [notesLocalLength]);
+
+  return (
+    <LoadButton
+      onClick={() => fetchMore()}
```
Comment (lyudad) [inline src/components/LoadMoreButton/index.tsx]: onClick={fetchMore}

File: src/components/Notes/CreateNote/index.tsx [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    amOrPm = 'PM';
+  }
+
+  const formattedDate = `${month} ${date}, ${year} at ${hour
+    .toString()
+    .padStart(2, '0')}:${minute.toString().padStart(2, '0')} ${amOrPm}`;
+
+  if (!addNew) {
+    return null;
+  }
```
Comment (lyudad) [inline src/components/Notes/CreateNote/index.tsx]: move to custom hook

File: src/components/Notes/Note/index.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+      )}
+      <Stack direction="row" alignItems="center" gap="30px">
+        <Doctor>
+          {t('Notes.dr')} {doctor.firstName} {doctor.lastName}
+        </Doctor>
+        <Stack
+          direction="row"
+          gap="5px"
+          alignItems="center"
+          onClick={() => handleDownload()}
```
Comment (lyudad) [inline src/components/Notes/Note/index.tsx]: onClick={handleDownload}

File: src/components/Notes/SearchBar/index.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+
+  const handleContainerClick = () => {
+    if (inputRef.current) {
+      inputRef.current.focus();
+    }
+  };
+
+  const [value, setValue] = React.useState<string>('');
```
Comment (lyudad) [inline src/components/Notes/SearchBar/index.tsx]: mess, there should be some order, for example: state, hooks, handlers, don't code like that

File: src/components/Notes/Skeleton/index.tsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import ContentLoader from 'react-content-loader';
+
+const Skeleton = () => (
+  <ContentLoader
+    speed={2}
+    viewBox="0 0 548 120"
+    backgroundColor="#f3f3f3"
```
Comment (lyudad) [inline src/components/Notes/Skeleton/index.tsx]: checklist

File: src/components/Notes/Sort/index.tsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        setToggleSortOrder(false);
+      }
+    };
+
+    document.addEventListener('mousedown', handleOutsideClick);
+
+    return () => {
+      document.removeEventListener('mousedown', handleOutsideClick);
+    };
+  }, []);
```
Comment (lyudad) [inline src/components/Notes/Sort/index.tsx]: move to custom hook

File: src/components/FileInput/styles.ts (hunk 2 above)
Comment (DaniilVolobuiev) [reply src/components/FileInput/styles.ts]: fixed

File: src/components/FileInput/styles.ts (hunk 3 above)
Comment (DaniilVolobuiev) [reply src/components/FileInput/styles.ts]: fixed

File: src/components/LoadMoreButton/index.tsx (hunk 4 above)
Comment (DaniilVolobuiev) [reply src/components/LoadMoreButton/index.tsx]: fixed

File: src/components/Notes/CreateNote/index.tsx (hunk 5 above)
Comment (DaniilVolobuiev) [reply src/components/Notes/CreateNote/index.tsx]: moved

File: src/components/Notes/Note/index.tsx (hunk 6 above)
Comment (DaniilVolobuiev) [reply src/components/Notes/Note/index.tsx]: fixed

File: src/components/Notes/Skeleton/index.tsx (hunk 8 above)
Comment (DaniilVolobuiev) [reply src/components/Notes/Skeleton/index.tsx]: fixed

File: src/components/Notes/Sort/index.tsx (hunk 9 above)
Comment (DaniilVolobuiev) [reply src/components/Notes/Sort/index.tsx]: moved

File: src/components/Notes/SearchBar/index.tsx (hunk 7 above)
Comment (DaniilVolobuiev) [reply src/components/Notes/SearchBar/index.tsx]: fixed

File: src/components/FileInput/index.tsx (hunk 1 above)
Comment (DaniilVolobuiev) [reply src/components/FileInput/index.tsx]: moved these lines to separate function
 let previewFiles: { [key: string]: any } = { ...files };
 let file = previewFiles[fileName];
 let isImageFile = file.type.split('/')[0] === 'image';
am i right?

File: src/components/FileInput/index.tsx [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    setFiles(toRemoveFiles);
+    callUpdateFilesCb(toRemoveFiles);
+  };
+
+  const getFileData = (files: { [key: string]: any }, fileName: string) => {
+    let previewFiles: { [key: string]: any } = { ...files };
+    let file = previewFiles[fileName];
+    let isImageFile = file.type.split('/')[0] === 'image';
+    return { file, isImageFile };
+  };
```
Comment (lyudad) [inline src/components/FileInput/index.tsx]: move logic t ocustom hook

File: src/components/Notes/AddNoteButton/styles.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+import FONT_ROBOTO from '@constants/fonts';
+
+export const Container = styled.div`
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  font-family: ${FONT_ROBOTO};
+  background-color: ${GHOST_WHITE};
+  height: 56px;
+  border: 1.5px solid #085dd7;
```
Comment (lyudad) [inline src/components/Notes/AddNoteButton/styles.ts]: ?

File: src/components/Notes/CreateNote/index.tsx [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+  const { t } = useTranslation();
+
+  const dateObj = new Date();
+  const month = dateObj.toLocaleString('en-US', { month: 'long' });
+  const date = dateObj.getDate();
+  const year = dateObj.getFullYear();
+  let hour = dateObj.getHours();
+  const minute = dateObj.getMinutes();
+  let amOrPm = 'AM';
```
Comment (lyudad) [inline src/components/Notes/CreateNote/index.tsx]: nonono what is it, it will be recalculated with each rerender, find other way

**Pull request merged**